projects/healthPingValidation.kp/knowledge.md

```diff
@@ -0,0 +1,550 @@
+        return ("Ping must countain at least one of the failures", ping)
+
+    
+    # validate sendFailure
+    supportedFailureTypes = ["eOK", "eRequest", "eUnreachable", "eChannelOpen", "eRedirect", "abort", "timeout"]
+    if sendFailure != None and len(sendFailure) > len(supportedFailureTypes):
+        return ("send Failure accept only 8 send failures", ping)
+    
+    if sendFailure != None:
+        for key in sendFailure.keys():
+            if not key in supportedFailureTypes:
+                return (key + " type is not supported", ping)
+        for count in sendFailure.values():
+            if not isinstance(count, int):
+                return ("Count must be int type", ping)
+        if sum(sendFailure.values()) > 60:
+            return ("sendFailure count must not be more than 60", ping)
+
+    
+     # validate pingDiscardedForSize
+    if pingDiscardedForSize != None:
+        if len(pingDiscardedForSize) > 10:
+            return ("pingDicardedForSize accept only top ten pings types", ping)
+        for count in pingDiscardedForSize.values():
+            if not isinstance(count, int):
+                return ("Count must be int type", ping)
+    
+    return (OK, ping)
+
+# retrieve all needed fields 
+validatedData = cachedData.map(validate)   
+errorsPerProblem = validatedData.countByKey()   
+errorsPerProblem
+```
+
+
+
+
+    defaultdict(int,
+                {'': 2013636, 'sendFailure count must not be more than 60': 3574})
+
+
+
+# Nightly 56 & 57. Investigate errors
+
+
+```python
+def printOSReason(data):
+    return "os: " + str(data[0]) + " reason: " + str(data[1])
+
+def osAndReasonForErros(error):
+    result = validatedData.filter(lambda pair: pair[0] == error).map(lambda pair: (pair[1]["payload/os"], pair[1]["payload/reason"])).collect()
+    return result[:min(10, len(result))]
+  
+print "Show only 10 info lines per problem \n"
+for err in errorsPerProblem.keys():
+    if err != '':
+        print err
+        print "\n".join(map(printOSReason, osAndReasonForErros(err)))
+        print "\n"
+```
+    Show only 10 info lines per problem 
+    
+    sendFailure count must not be more than 60
+    os: {u'version': u'10.0', u'name': u'WINNT'} reason: immediate
+    os: {u'version': u'6.1', u'name': u'WINNT'} reason: immediate
+    os: {u'version': u'3.10.65', u'name': u'Android'} reason: delayed
+    os: {u'version': u'4.9.0-kali1-amd64', u'name': u'Linux'} reason: shutdown
+    os: {u'version': u'10.0', u'name': u'WINNT'} reason: immediate
+    os: {u'version': u'10.0', u'name': u'WINNT'} reason: shutdown
+    os: {u'version': u'10.0', u'name': u'WINNT'} reason: shutdown
+    os: {u'version': u'10.0', u'name': u'WINNT'} reason: shutdown
+    os: {u'version': u'10.0', u'name': u'WINNT'} reason: immediate
+    os: {u'version': u'10.0', u'name': u'WINNT'} reason: shutdown
+    
+    
+
+
+# Nightly 56 & 57. Compute pings count per day
+This include showing diagrams and printings stats
```

Comment:
typo "include" should be "includes" and "printings" should be "printing" (English is an awful, inconsistent language)